jrt/src/com/yahoo/jrt/TlsCryptoSocket.java

```diff
@@ -0,0 +1,254 @@
+    public FlushResult flush() throws IOException {
+        channelWrite();
+        return hasWrapBufferMoreData() ? FlushResult.NEED_WRITE : FlushResult.DONE;
+    }
+
+    private boolean handshakeWrap() throws IOException {
+        SSLEngineResult result = sslEngineWrap(NULL_BUFFER);
+        switch (result.getStatus()) {
+            case OK:
+                return true;
+            case BUFFER_OVERFLOW:
+                return false;
+            default:
+                throw unexpectedStatusException(result.getStatus());
+        }
+    }
+
+    private int applicationDataWrap(ByteBuffer src) throws IOException {
+        SSLEngineResult result = sslEngineWrap(src);
+        switch (result.getStatus()) {
+            case OK:
+                int bytesConsumed = result.bytesConsumed();
+                if (bytesConsumed == 0) throw new SSLException("Got handshake data in application data wrap");
+                return bytesConsumed;
+            case BUFFER_OVERFLOW:
+                return 0;
+            default:
+                throw unexpectedStatusException(result.getStatus());
+        }
+    }
+
+    private SSLEngineResult sslEngineWrap(ByteBuffer src) throws IOException {
+        SSLEngineResult result = sslEngine.wrap(src, wrapBuffer.getWritable(0));
+        if (result.getStatus() == Status.CLOSED) throw new ClosedChannelException();
+        return result;
+    }
+
+    private boolean handshakeUnwrap() throws IOException {
+        SSLEngineResult result = sslEngineUnwrap(handshakeDummyBuffer);
+        switch (result.getStatus()) {
+            case OK:
+                if (result.bytesProduced() > 0) throw new SSLException("Got application data in handshake unwrap");
+                return true;
+            case BUFFER_UNDERFLOW:
+                return false;
+            default:
+                throw unexpectedStatusException(result.getStatus());
+        }
+    }
+
+    private int applicationDataUnwrap(ByteBuffer dst) throws IOException {
+        SSLEngineResult result = sslEngineUnwrap(dst);
+        switch (result.getStatus()) {
+            case OK:
+                int bytesProduced = result.bytesProduced();
+                if (bytesProduced == 0) throw new SSLException("Got handshake data in application data unwrap");
+                return bytesProduced;
+            case BUFFER_OVERFLOW:
+            case BUFFER_UNDERFLOW:
+                return 0;
+            default:
+                throw unexpectedStatusException(result.getStatus());
+        }
+    }
+
+    private SSLEngineResult sslEngineUnwrap(ByteBuffer dst) throws IOException {
+        SSLEngineResult result = sslEngine.unwrap(unwrapBuffer.getReadable(), dst);
+        if (result.getStatus() == Status.CLOSED) throw new ClosedChannelException();
+        return result;
+    }
+
+    // returns number of bytes read
+    private int channelRead() throws IOException {
+        int read = channel.read(unwrapBuffer.getWritable(0));
```

Comment:
should give something larger than 0 as parameter to ensure we try to read a reasonable chunk of data